### DISCUSSION TITLE
ServiceFault from server received while waiting for publish response - 106

### DISCUSSION THREAD

**Discussion by saimanoj567** [FreeOpcUa/python-opcua#1300] [Q&A]
Hello All, I have an issue while deleting the subscription. I was able to unsubscribe. but I get error when I reach the point sub.delete().I have packed the handles in the list and looping it for unsubscribe. I also read few issues and made some changes based on the discussions. I'm using the library version - (0.98.13) and Server - Bechkoff . 

( I have also changed MaxNotificationsPerPublish from zero to 1000 ). 

Do I need to change any other parameters in server configuration ?

![wireshark](https://user-images.githubusercontent.com/50169720/112899029-e6411b80-90e1-11eb-8594-914d75ebdf67.PNG)
![error](https://user-images.githubusercontent.com/50169720/112899043-e9d4a280-90e1-11eb-8592-4c469e21e5d8.PNG)

**Comment by AndreasHeine** [comment]
try delete ? except ServiceFault!

sometimes its wise to make each request in a try except because each request could fail

**Reply by AndreasHeine** [reply]
could you export the relavant lines of the wireshark trace and post it here?

**Reply by saimanoj567** [reply]
Thank you for your response I tried to view the subscription and total no of monitered item in ua expext its shows exactly the same number of subscription nodes ( total node - 9). once I execute subcription delete the created subscription id gets deleted as well but i get the error message.

[logs.txt](https://github.com/FreeOpcUa/python-opcua/files/6227701/logs.txt)

![before](https://user-images.githubusercontent.com/50169720/112960786-0950e700-9145-11eb-9045-95dbdf10889a.PNG)

![after](https://user-images.githubusercontent.com/50169720/112960818-1372e580-9145-11eb-9932-9ebf30a2f614.PNG)

![after delete](https://user-images.githubusercontent.com/50169720/112960916-2d142d00-9145-11eb-967d-185591268cf8.PNG)

**Comment by saimanoj567** [comment]
Whether this error occurs on server side/ client side ?

**Reply by AndreasHeine** [reply]
the service fault comes from server and client raises because he's got a service fault as response.

**Reply by saimanoj567** [reply]
Thankyou Andreas  .Have you ever faced these issues with bechkoff systems before ? I have also attached logs does it helped ?

**Reply by AndreasHeine** [reply]
it should be no vendor specific topic in wireshark you can see who send what most peaple dont check the ServerCapabilities and OperationLimits up front 

tbh have never used Beckhoff... We have 99.9999% Siemens

maxNotificationsPerPublish:
The maximum number of notifications that the Client wishes to receive in a single Publish response. A value of zero indicates that there is no limit.
The number of notifications per Publish is the sum of monitoredItems in the DataChangeNotification and events in the EventNotificationList.

keep in mind that values can change a lot and if the server produces more notifications then the client consumes (depends on the queue size!) it could be that the server runs out of ressources.

**but at first you need to check the errorcode in the service fault to get a hint what causes the fault!**

**Reply by saimanoj567** [reply]
Default bechkoff SERVER CONFIG:
```
<MinPublishingInterval>50</MinPublishingInterval>
  <MaxPublishingInterval>0</MaxPublishingInterval>
  <MinKeepAliveInterval>5000</MinKeepAliveInterval>
  <MinSubscriptionLifetime>1000</MinSubscriptionLifetime>
  <MaxSubscriptionLifetime>0</MaxSubscriptionLifetime>
  <MaxRetransmitionQueueSize>10</MaxRetransmitionQueueSize>
  <MaxNotificationsPerPublish>0</MaxNotificationsPerPublish>
  <MaxDataQueueSize>100</MaxDataQueueSize>
  <MaxEventQueueSize>10000</MaxEventQueueSize>
  <MaxSubscriptionCount>0</MaxSubscriptionCount>
  <MaxMonitoredItemCount>0</MaxMonitoredItemCount>
  <MaxMonitoredItemPerSubscriptionCount>0</MaxMonitoredItemPerSubscriptionCount>

```

logs:
```
INFO:opcua.client.ua_client:delete_subscription
DEBUG:opcua.client.ua_client.Socket:Sending: DeleteSubscriptionsRequest(TypeId:{self.TypeId}, RequestHeader:{self.RequestHeader}, Parameters:{self.Parameters})
INFO:opcua.uaprotocol:received header: Header(type:b'MSG', chunk_type:b'F', body_size:40, channel:695954467)
INFO:opcua.client.ua_client:call_publish_callback
WARNING:opcua.client.ua_client.Socket:ServiceFault from server received while waiting for publish response
INFO:opcua.client.ua_client:BadNoSubscription received, ignoring because it's probably valid.
DEBUG:opcua.uaprotocol:Waiting for header
INFO:opcua.uaprotocol:received header: Header(type:b'MSG', chunk_type:b'F', body_size:40, channel:695954467)
INFO:opcua.client.ua_client:call_publish_callback
WARNING:opcua.client.ua_client.Socket:ServiceFault from server received while waiting for publish response
INFO:opcua.client.ua_client:BadNoSubscription received, ignoring because it's probably valid.
DEBUG:opcua.uaprotocol:Waiting for header
INFO:opcua.uaprotocol:received header: Header(type:b'MSG', chunk_type:b'F', body_size:52, channel:695954467)
INFO:opcua.client.ua_client:_delete_subscriptions_callback
DEBUG:opcua.client.ua_client:DeleteSubscriptionsResponse(TypeId:{self.TypeId}, ResponseHeader:{self.ResponseHeader}, Results:{self.Results}, DiagnosticInfos:{self.DiagnosticInfos})
DEBUG:opcua.uaprotocol:Waiting for header
DEBUG:opcua.client.client:stoping keepalive thread
DEBUG:opcua.client.client:keepalive thread has stopped
INFO:opcua.client.ua_client:close_session
DEBUG:opcua.client.ua_client.Socket:Sending: CloseSessionRequest(TypeId:{self.TypeId}, RequestHeader:{self.RequestHeader}, DeleteSubscriptions:{self.DeleteSubscriptions})
INFO:opcua.uaprotocol:received header: Header(type:b'MSG', chunk_type:b'F', body_size:40, channel:695954467)
DEBUG:opcua.uaprotocol:Waiting for header
INFO:opcua.client.ua_client.Socket:close_secure_channel
DEBUG:opcua.client.ua_client.Socket:Sending: CloseSecureChannelRequest(TypeId:{self.TypeId}, RequestHeader:{self.RequestHeader})
INFO:opcua.client.ua_client.Socket:Request to close socket received
INFO:opcua.client.ua_client.Socket:Socket closed, waiting for receiver thread to terminate...
INFO:opcua.client.ua_client.Socket:Socket has closed connection
INFO:opcua.client.ua_client.Socket:Cancelling request 32
INFO:opcua.client.ua_client.Socket:Thread ended
INFO:opcua.client.ua_client.Socket:Done closing socket: Receiving thread terminated, socket disconnected
```
